examples/post-sync/controller.go

```diff
@@ -0,0 +1,312 @@
+	// Validate sync kind
+	if syncKind != RootSyncKind && syncKind != RepoSyncKind {
+		panic(fmt.Sprintf("invalid sync kind: %s, must be either %s or %s", syncKind, RootSyncKind, RepoSyncKind))
+	}
+	return &SyncStatusController{
+		client:        client,
+		log:           log,
+		statusTracker: statusTracker,
+		syncKind:      syncKind,
+	}
+}
+
+// Reconcile handles RootSync and RepoSync resources
+func (c *SyncStatusController) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	var syncID SyncID
+	var status v1beta1.Status
+	var generation int64
+	var observedGeneration int64
+	var errs, commit string
+
+	// Use the explicit syncKind to know which type to fetch
+	switch c.syncKind {
+	case RootSyncKind:
+		var root v1beta1.RootSync
+		if err := c.client.Get(ctx, req.NamespacedName, &root); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+		syncID = c.getSyncID(ctx, req)
+		status = root.Status.Status
+		generation = root.Generation
+		observedGeneration = root.Status.ObservedGeneration
+		errs, commit = c.processRootSyncConditions(root.Status.Conditions)
+
+	case RepoSyncKind:
+		var repo v1beta1.RepoSync
+		if err := c.client.Get(ctx, req.NamespacedName, &repo); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+		syncID = c.getSyncID(ctx, req)
+		status = repo.Status.Status
+		generation = repo.Generation
+		observedGeneration = repo.Status.ObservedGeneration
+		errs, commit = c.processRepoSyncConditions(repo.Status.Conditions)
+
+	default:
+		c.log.Error(fmt.Errorf("unknown sync kind: %s", c.syncKind), "unrecognized syncKind")
+		return reconcile.Result{}, fmt.Errorf("unknown sync kind: %s", c.syncKind)
+	}
+
+	// Process any errors from conditions
+	if err := c.handleErrors(syncID, errs, commit, generation, observedGeneration, false); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Process sync status
+	if !isSuccessfulSync(status, generation) {
+		errMessage, commit, isTruncated := extractErrorAndCommitFromStatus(status)
+		if err := c.handleErrors(syncID, errMessage, commit, generation, observedGeneration, isTruncated); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *SyncStatusController) getSyncID(ctx context.Context, req reconcile.Request) SyncID {
+	return SyncID{
+		Name:      req.Name,
+		Kind:      c.syncKind,
+		Namespace: req.Namespace,
+	}
+}
+
+// handleErrors aggregates error messages and logs them if they haven't been logged already
+func (c *SyncStatusController) handleErrors(rsyncID SyncID, errMessage string, commit string, generation int64, observedGeneration int64, isTruncated bool) error {
+	if errMessage == "" {
+		return nil
+	}
+	if c.statusTracker.IsLogged(rsyncID, commit, errMessage) {
+		return nil
+	}
+	c.log.Info("Sync error detected",
+		"sync", rsyncID.Name,
+		"namespace", rsyncID.Namespace,
+		"kind", rsyncID.Kind,
```

Comment:
does the order matter here? if not, i'd put it in order of significance: kind, sync, namespace, name, etc.

Reply:
The order doesn't matter much here as the log is getting recorded as json, so that later the cloud function would grab them by key. The final representation of the error log is determined at stage like cloud function or other integration level when pushing the message to user. Still modified for readability.